File: backend/dashboard_api.py
"""
Enhanced Dashboard API for SavvyCleanse
Provides chart data and analytics endpoints for the dashboard
"""

from fastapi import APIRouter, HTTPException, Depends, Query
from typing import Dict, Any, List, Optional
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
import json

router = APIRouter(prefix="/dashboard", tags=["dashboard"])

# Mock data generators for demonstration
def generate_sales_performance_data(days: int = 365) -> List[Dict[str, Any]]:
    """Generate mock sales performance data"""
    months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 
              'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
    
    data = []
    base_sales = 4000
    base_target = 3800
    
    for i, month in enumerate(months):
        # Add seasonal variation and growth trend
        seasonal_factor = 1 + 0.2 * np.sin(i * np.pi / 6)
        growth_factor = 1 + (i * 0.05)
        noise = np.random.normal(0, 200)
        
        sales = int(base_sales * seasonal_factor * growth_factor + noise)
        target = int(base_target * growth_factor)
        
        data.append({
            'month': month,
            'sales': max(sales, 0),
            'target': target
        })
    
    return data

def generate_market_share_data() -> List[Dict[str, Any]]:
    """Generate mock market share data"""
    products = [
        {'name': 'Product A', 'value': 4500},
        {'name': 'Product B', 'value': 3200},
        {'name': 'Product C', 'value': 2600},
        {'name': 'Product D', 'value': 1900},
        {'name': 'Others', 'value': 650}
    ]
    
    total_value = sum(p['value'] for p in products)
    
    for product in products:
        product['percentage'] = round((product['value'] / total_value) * 100, 1)
    
    return products

def generate_user_engagement_data(weeks: int = 8) -> List[Dict[str, Any]]:
    """Generate mock user engagement data"""
    data = []
    base_active = 1200
    base_new = 180
    
    for i in range(weeks):
        # Add weekly variation
        active_variation = np.random.normal(0, 100)
        new_variation = np.random.normal(0, 30)
        
        # Add growth trend
        growth_factor = 1 + (i * 0.02)
        
        active_users = int(base_active * growth_factor + active_variation)
        new_users = int(base_new * growth_factor + new_variation)
        
        data.append({
            'week': f'Week {i + 1}',
            'active_users': max(active_users, 0),
            'new_users': max(new_users, 0)
        })
    
    return data

def generate_quarterly_performance_data() -> List[Dict[str, Any]]:
    """Generate mock quarterly performance data"""
    quarters = ['Q1 2023', 'Q2 2023', 'Q3 2023', 'Q4 2023', 'Q1 2024', 'Q2 2024']
    base_revenue = 125000
    
    data = []
    for i, quarter in enumerate(quarters):
        # Add quarterly growth
        growth_factor = 1 + (i * 0.12)
        seasonal_factor = 1 + (0.1 if 'Q4' in quarter else 0)
        
        revenue = int(base_revenue * growth_factor * seasonal_factor)
        growth_percentage = round(((revenue / base_revenue) - 1) * 100, 1)
        
        data.append({
            'quarter': quarter,
            'revenue': revenue,
            'growth': growth_percentage
        })
    
    return data

def apply_analysis_mode(data: Dict[str, Any], mode: str) -> Dict[str, Any]:
    """Apply analysis mode transformations to the data"""
    if mode == 'descriptive':
        # Return data as-is for descriptive analysis
        return data
    
    elif mode == 'diagnostic':
        # Add correlation and relationship insights
        data['insights'] = {
            'correlations': [
                {'variables': ['sales', 'marketing_spend'], 'correlation': 0.78, 'p_value': 0.001},
                {'variables': ['user_engagement', 'revenue'], 'correlation': 0.65, 'p_value': 0.02}
            ],
            'key_findings': [
                'Strong positive correlation between marketing spend and sales',
                'User engagement significantly impacts revenue generation',
                'Seasonal patterns detected in Q4 performance'
            ]
        }
    
    elif mode == 'predictive':
        # Add forecast data and predictions
        data['forecasts'] = {
            'next_quarter_revenue': 245000,
            'confidence_interval': [220000, 270000],
            'growth_prediction': 15.2,
            'key_drivers': [
                {'factor': 'User Engagement', 'impact': 0.3},
                {'factor': 'Market Expansion', 'impact': 0.2},
                {'factor': 'Product Innovation', 'impact': 0.15}
            ]
        }
    
    elif mode == 'prescriptive':
        # Add optimization recommendations
        data['recommendations'] = {
            'actions': [
                {
                    'action': 'Increase marketing spend by 20% in Q2',
                    'expected_impact': '+$340K revenue',
                    'confidence': 0.85,
                    'priority': 'High'
                },
                {
                    'action': 'Focus Product A expansion in Europe',
                    'expected_impact': '+$180K revenue',
                    'confidence': 0.72,
                    'priority': 'Medium'
                },
                {
                    'action': 'Optimize user onboarding flow',
                    'expected_impact': '+12% conversion rate',
                    'confidence': 0.68,
                    'priority': 'Medium'
                }
            ],
            'optimization_score': 8.7
        }
    
    return data

@router.get("/data")
async def get_dashboard_data(
    analysis_mode: str = Query(default="descriptive", description="Analysis mode: descriptive, diagnostic, predictive, prescriptive"),
    date_range: str = Query(default="last_30_days", description="Date range filter"),
    region: str = Query(default="all", description="Region filter"),
    category: str = Query(default="all", description="Category filter")
):
    """Get dashboard chart data with applied filters and analysis mode"""
    try:
        # Generate base data
        dashboard_data = {
            'sales_performance': generate_sales_performance_data(),
            'market_share': generate_market_share_data(),
            'user_engagement': generate_user_engagement_data(),
            'quarterly_performance': generate_quarterly_performance_data()
        }
        
        # Apply filters (simplified for demo)
        if date_range != "all":
            # In a real implementation, this would filter based on actual dates
            pass
        
        if region != "all":
            # In a real implementation, this would filter by region
            pass
        
        if category != "all":
            # In a real implementation, this would filter by category
            pass
        
        # Apply analysis mode transformations
        enhanced_data = apply_analysis_mode(dashboard_data, analysis_mode)
        
        # Add metadata
        enhanced_data['metadata'] = {
            'generated_at': datetime.now().isoformat(),
            'analysis_mode': analysis_mode,
            'filters_applied': {
                'date_range': date_range,
                'region': region,
                'category': category
            },
            'data_quality_score': 9.2,
            'last_updated': datetime.now().isoformat()
        }
        
        return enhanced_data
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error generating dashboard data: {str(e)}")

@router.post("/analyze/{analysis_type}")
async def run_dashboard_analysis(
    analysis_type: str,
    dataset_id: Optional[str] = None,
    target_column: Optional[str] = None,
    filters: Optional[Dict[str, Any]] = None
):
    """Run specific analysis on dashboard data"""
    try:
        if analysis_type not in ['descriptive', 'diagnostic', 'predictive', 'prescriptive']:
            raise HTTPException(status_code=400, detail="Invalid analysis type")
        
        # Get base dashboard data
        dashboard_data = {
            'sales_performance': generate_sales_performance_data(),
            'market_share': generate_market_share_data(),
            'user_engagement': generate_user_engagement_data(),
            'quarterly_performance': generate_quarterly_performance_data()
        }
        
        # Apply analysis
        result = apply_analysis_mode(dashboard_data, analysis_type)
        
        # Add analysis-specific metadata
        result['analysis_metadata'] = {
            'type': analysis_type,
            'dataset_id': dataset_id,
            'target_column': target_column,
            'executed_at': datetime.now().isoformat(),
            'processing_time_ms': np.random.randint(200, 800)  # Simulate processing time
        }
        
        return result
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error running analysis: {str(e)}")

@router.get("/summary")
async def get_dashboard_summary():
    """Get dashboard summary statistics"""
    try:
        return {
            'total_datasets': 15,
            'total_analyses': 47,
            'active_dashboards': 8,
            'data_quality_avg': 8.9,
            'last_updated': datetime.now().isoformat(),
            'system_status': 'healthy',
            'processing_queue': 2
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error getting summary: {str(e)}")

@router.post("/export")
async def export_dashboard_data(
    format: str = Query(default="json", description="Export format: json, csv, excel"),
    charts: Optional[List[str]] = Query(default=None, description="Specific charts to export")
):
    """Export dashboard data in specified format"""
    try:
        # Get current dashboard data
        dashboard_data = {
            'sales_performance': generate_sales_performance_data(),
            'market_share': generate_market_share_data(),
            'user_engagement': generate_user_engagement_data(),
            'quarterly_performance': generate_quarterly_performance_data()
        }
        
        if charts:
            # Filter to only requested charts
            filtered_data = {chart: dashboard_data.get(chart, []) for chart in charts if chart in dashboard_data}
            dashboard_data = filtered_data
        
        if format == "json":
            return dashboard_data
        elif format == "csv":
            # Convert to CSV format (simplified)
            return {"message": "CSV export functionality would be implemented here"}
        elif format == "excel":
            # Convert to Excel format (simplified)
            return {"message": "Excel export functionality would be implemented here"}
        else:
            raise HTTPException(status_code=400, detail="Unsupported export format")
            
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error exporting data: {str(e)}")

File: src/utils/dashboardApi.ts
/**
 * Dashboard API utilities for fetching chart data and analytics
 */

export interface DashboardData {
  sales_performance: Array<{
    month: string;
    sales: number;
    target: number;
  }>;
  market_share: Array<{
    name: string;
    value: number;
    percentage: number;
  }>;
  user_engagement: Array<{
    week: string;
    active_users: number;
    new_users: number;
  }>;
  quarterly_performance: Array<{
    quarter: string;
    revenue: number;
    growth: number;
  }>;
  metadata?: {
    generated_at: string;
    analysis_mode: string;
    filters_applied: Record<string, string>;
    data_quality_score: number;
    last_updated: string;
  };
  insights?: any;
  forecasts?: any;
  recommendations?: any;
}

export interface DashboardFilterOptions {
  analysis_mode?: string;
  date_range?: string;
  region?: string;
  category?: string;
}

const generateMockData = (filters: DashboardFilterOptions = {}): DashboardData => {
  const mode = filters.analysis_mode || 'descriptive';
  
  return {
    sales_performance: [
      { month: 'Jan', sales: 4200, target: 4000 },
      { month: 'Feb', sales: 3800, target: 4100 },
      { month: 'Mar', sales: 5100, target: 4200 },
      { month: 'Apr', sales: 4600, target: 4300 },
      { month: 'May', sales: 5400, target: 4400 },
      { month: 'Jun', sales: 4900, target: 4500 }
    ],
    market_share: [
      { name: 'Product A', value: 4500, percentage: 35 },
      { name: 'Product B', value: 3200, percentage: 25 },
      { name: 'Product C', value: 2600, percentage: 20 },
      { name: 'Product D', value: 1900, percentage: 15 },
      { name: 'Others', value: 650, percentage: 5 }
    ],
    user_engagement: [
      { week: 'Week 1', active_users: 1200, new_users: 180 },
      { week: 'Week 2', active_users: 1350, new_users: 220 },
      { week: 'Week 3', active_users: 1180, new_users: 160 },
      { week: 'Week 4', active_users: 1420, new_users: 280 }
    ],
    quarterly_performance: [
      { quarter: 'Q1 2023', revenue: 125000, growth: 12 },
      { quarter: 'Q2 2023', revenue: 142000, growth: 18 },
      { quarter: 'Q3 2023', revenue: 158000, growth: 22 },
      { quarter: 'Q4 2023', revenue: 175000, growth: 28 }
    ],
    metadata: {
      generated_at: new Date().toISOString(),
      analysis_mode: mode,
      filters_applied: filters as Record<string, string>,
      data_quality_score: 8.9,
      last_updated: new Date().toISOString()
    }
  };
};

export const fetchDashboardData = async (filters: DashboardFilterOptions = {}): Promise<DashboardData> => {
  // Return local mock data (no backend server available)
  return generateMockData(filters);
};

export const runDashboardAnalysis = async (
  analysisType: string,
  _datasetId?: string,
  _targetColumn?: string,
  _filters?: Record<string, any>
): Promise<any> => {
  // Mock response since no backend is available
  return {
    analysis_type: analysisType,
    status: 'completed',
    results: {
      summary: `${analysisType} analysis completed successfully`,
      metrics: { accuracy: 0.92, confidence: 0.87 }
    }
  };
};

export const exportDashboardData = async (
  format: string = 'json',
  _charts?: string[]
): Promise<any> => {
  // Mock export
  return {
    status: 'success',
    format,
    download_url: '#',
    message: 'Export generated successfully'
  };
};

export const getDashboardSummary = async (): Promise<any> => {
  return {
    total_datasets: 15,
    total_analyses: 47,
    active_dashboards: 8,
    data_quality_avg: 8.9,
    system_status: 'healthy'
  };
};